var/spack/repos/builtin/packages/hpx/package.py

```diff
@@ -69,6 +78,12 @@ class Hpx(CMakePackage, CudaPackage):
     # https://github.com/STEllAR-GROUP/hpx/issues/4728#issuecomment-640685308
     depends_on('boost@:1.72.0', when='@:1.4')
 
+    # COROUTINES
+    # https://github.com/STEllAR-GROUP/hpx/issues/4829
+    depends_on('boost+context', when='+generic_coroutines')
+    _msg_generic_coroutines = 'This platform requires +generic_coroutines'
+    conflicts('~generic_coroutines', when='platform=darwin', msg=_msg_generic_coroutines)
```

Comment:
Shouldn't this rather use the inverted condition `sys.platform.startswith('linux') or sys.platform == 'win32'` rather than checking for Darwin (I don't know what the correct incantation in this context would be)?

Reply:
From https://github.com/STEllAR-GROUP/hpx/issues/4829#issuecomment-660919450
> I was not able to figure out a way to get the dual constraint (e.g. "all platforms except linux and windows"), so once we know the additional platforms that do not support generic_coroutines, we should extend the spack package copying what it has been done for darwin.

So, I'm with you! I thought about this and I may have a sort of workaround. Usually the `when` clause is used, that checks the spec to match the criteria; but AFAIK the spec language does not allow to state "not-match", but just "match".
In this specific case, the workaround may be to replace the `conflicts(when=)` with something like:

```python
if not (sys.platform.startswith('linux') or sys.platform == 'win32'):
  conflicts('~generic_coroutines')
```

I'm going to ask to spack main developers if it is a good way to do this ;)

Reply:
They gave me a very useful answer!

Currently in `spack` the supported platforms are `linux`, `darwin`, `cray`, `bgq`. From this, it is possible to state that we can just don't care about `win32` (and even less about `cygwin`).

Moreover, they suggested me this snippet: it allows us to explicitly state that a platform is supported by HPX custom coroutines (allowing the `~generic_coroutines` variant), otherwise it ensures that any other platform that spack can think of in the future will work just with `+generic_coroutines`.

```python
import spack.architecture

_supported_platforms = ['linux']
for platform in spack.architecture.all_platforms():
    if platform.name not in (_supported_platforms):
        conflicts('~generic_coroutines', when='platform=%s' % platform.name)
```

Is it correct to say that HPX custom coroutines are supported also by `cray`?

Reply:
> Is it correct to say that HPX custom coroutines are supported also by cray?

It depends on the architecture. If 'cray' means any of the old Cray machines - I don't know the answer. If 'cray' means one of the modern Cray machines, then we probably support it as those are x86 based, afaik.

Reply:
I'm pretty sure we can extend the check to target architecture...but at this point it is not clear to me what it is the constraint for using the POSIX coroutines (`~generic_coroutines`).

https://github.com/STEllAR-GROUP/hpx/blob/82a488342d583166f39739f31fb66c53dd6df659/CMakeLists.txt#L1672-L1681
```cmake
# ##############################################################################
# Set configuration option to use Boost.Context or not. This depends on the
# platform.
set(__use_generic_coroutine_context OFF)
if(APPLE)
  set(__use_generic_coroutine_context ON)
endif()
if("${HPX_PLATFORM_UC}" STREQUAL "BLUEGENEQ")
  set(__use_generic_coroutine_context ON)
endif()
```

From the above snippet the only constraint I see are about `APPLE` and `BLUEGENEQ` that explicitly requires `+generic_coroutines`. So from this, I would just try finding a way to exclude these two in spack (i.e. `darwin` and `bgq`).

I know I'm deriving the constraint from the CMake scripts, and they may not cover all cases. Trying to start getting an overview, can you correct and extend this?

- `linux` is supported generically, and it does not need any architecture check (from [HPX CMake ARM toolchain](https://github.com/STEllAR-GROUP/hpx/blob/master/cmake/toolchains/ARM-gcc.cmake) generic coroutines are enabled, but it looks like a preference and not a constraint);
- `cray` it is equipped with linux, but from you last comment it seems the compatibility with POSIX coroutines may depend on the architecture (just `x86_64`?)
- `bgq` it is equipped with linux, but from the above snippet it does not support POSIX coroutines.
- `darwin` is unix-based, and even if my machine is `x86_64`, it is not compatible with POSIX coroutines;

In addition to these, [HPX CMake BGION toolchain](https://github.com/STEllAR-GROUP/hpx/blob/82a488342d583166f39739f31fb66c53dd6df659/cmake/toolchains/BGION-gcc.cmake#L54-L58) exists, it is reported as linux (is it [this](https://www.mcs.anl.gov/~kazutomo/ion-kernel-boot.html) `powerpc` architecture?) and it looks like it prefers POSIX coroutines (preference, not constraint).

Do you know exactly what the constraint is? Is it a mix of OS/architecture/... ?

Reply:
Let's not overcomplicate things. Going with `!Darwin` is good enough for now. Let's cross the other bridges once we have to.